util/topgen.py

```diff
@@ -569,6 +569,56 @@ def generate_clkmgr(top, cfg_path, out_path):
     gen_rtl.gen_rtl(hjson_obj, str(rtl_path))
 
 
+# generate pwrmgr
+def generate_pwrmgr(top, out_path):
+    log.info("Generating pwrmgr")
+
+    # Count number of interrupts
+    n_wkups = len(top["wkups"])
+    log.info("Found {} wakeup signals".format(n_wkups))
+
+    if n_wkups < 1:
+        n_wkups = 1
+        log.warning("The design has no wakeup sources. Low power not supported")
+
+    # Define target path
+    rtl_path = out_path / 'ip/pwrmgr/rtl/autogen'
+    rtl_path.mkdir(parents=True, exist_ok=True)
+    doc_path = out_path / 'ip/pwrmgr/data/autogen'
+    doc_path.mkdir(parents=True, exist_ok=True)
+
+    # So, read template files from ip directory.
+    tpl_path = out_path / '../ip/pwrmgr/data'
+    hjson_tpl_path = tpl_path / 'pwrmgr.hjson.tpl'
+
+    # Render and write out hjson
+    out = StringIO()
```

Comment:
beginning to think we should make this code into lib and just invoke it... see quite a bit of repeat here..

Reply:
Good Idea. Look at the `generate_top` function in topgen. You can re-use it. temporary